hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/FsDatasetImpl.java

```diff
@@ -2350,6 +2350,9 @@ private void invalidate(String bpid, Block[] invalidBlks, boolean async)
         removing = volumeMap.remove(bpid, invalidBlks[i]);
         addDeletingBlock(bpid, removing.getBlockId());
         LOG.debug("Block file {} is to be deleted", removing.getBlockURI());
+        if (datanode.getMetrics() != null) {
```

Comment:
Ah, I believe that this will make `BlocksRemoved` more precisions. But I do not think it will be exact, consider that `removing == null` which means that DataNode do not manage this pending delete replica, so we should not increase the metric. Maybe there are some other cases, not thought deeply. I just suggest we should consider if removing will be null.
Another point, line 2352 will throw NPE but not decide if removing will null or not. This is another issue, not related to this PR.

Reply:
@Hexiaoqiao Thanks for your review. I learned a lot and have updated this patch, please help me review it again.

Reply:
@ZanderXu Copy, sir. 

> Another one, it is unnecessary to decide if datanode.getMetrics() is null or not

We need it, because `hadoop.hdfs.server.datanode.fsdataset.impl.TestFsDatasetImpl` try to invalidate some blocks with null metrics. Or maybe we can modify the UT.

> "Wish every PR could be merged without any issues reviewers feedback." 

+1, and I will do my best to achieve it.

Reply:
> We need it, because hadoop.hdfs.server.datanode.fsdataset.impl.TestFsDatasetImpl try to invalidate some blocks with null metrics.

I found that there are many `datanode.getMetrics()` calls without any null check in FsDatasetImpl, this is the same one. Not sure where will throw NPE, if that we should file another PR to fix them? Thanks.

Reply:
@Hexiaoqiao I have updated this code and fixed the failed UT `org.apache.hadoop.hdfs.server.datanode.fsdataset.impl.TestFsDatasetImpl`, please help me review it again. Thanks